lib/dockerregistry/inventory.go

```diff
@@ -2527,42 +2527,121 @@ func GetDeleteCmd(
 		cmd)
 }
 
-// Contains checks whether a given Manifest mentions the contents of a
-// gcrPayload (re-interpreted as a FQIN or PQIN).
-// nolint[gocyclo]
-func (m Manifest) Contains(gcrPayload GCRPubSubPayload) bool {
-	fqinOnly := len(gcrPayload.Tag) == 0
+// GcrPayloadMatch holds bitwise flags for matching a GCRPubSubPayload against a
+// promoter manifest.
+//
+// For an example of how bitwise flags work, see
+// https://yourbasic.org/golang/bitmask-flag-set-clear/.
+type GcrPayloadMatch uint
+
+const (
+	// FlagPathMatched is set if the registry + image path (everything leading
+	// up to either the digest or a tag) matches.
+	FlagPathMatched GcrPayloadMatch = 1 << iota
+	// FlagDigestMatched is set if the digest in the payload matches a digest in
+	// the promoter manifest. This is ONLY matched if the path also matches.
+	FlagDigestMatched
+	// FlagTagMatched is ONLY matched if the digest also matches.
+	FlagTagMatched
+	// FlagTagMismatch is only set if the digest matches, but the tag found in
+	// the payload does NOT match what is found in the promoter manifest for the
+	// digest. This can happen if somone manually tweaks a tag in GCR (assume
+	// bad actor) to something other than what is specified in the promoter
+	// manifest.
+	FlagTagMismatch
+)
 
+// Match checks whether a GCRPubSubPayload is mentioned in a Manifest. The
+// degree of the match is reflected in the GcrPayloadMatch result.
+func (payload GCRPubSubPayload) Match(m Manifest) GcrPayloadMatch {
+	var r GcrPayloadMatch
 	for _, rc := range m.Registries {
-		if rc.Src {
+		r = payload.matchImages(rc, m.Images)
+		if r&FlagPathMatched != 0 {
+			return r
+		}
+	}
+	return r
+}
+
+func (payload GCRPubSubPayload) matchImages(
+	rc RegistryContext,
+	images []Image,
+) GcrPayloadMatch {
+
+	var r GcrPayloadMatch
+	// We do not look at source registries, because the payload will only
+	// contain the image name as it appears on the destination (production).
+	// So the prefix and substring checks only make sense for the
+	// destination registries.
+	if rc.Src {
+		return r
+	}
+	// Speed up the search by skipping over registry names whose leading
+	// characters do not match.
+	if !strings.HasPrefix(payload.Digest, (string)(rc.Name)) {
+		return r
+	}
+
+	for _, image := range images {
+		r = payload.matchImage(rc, image)
+		// If we have just a path match, return early because we should
+		// limit the scope of the search to just 1 image.
+		if r&FlagPathMatched != 0 {
+			return r
+		}
+	}
+	return r
+}
+
+func (payload GCRPubSubPayload) matchImage(
+	rc RegistryContext,
+	image Image,
+) GcrPayloadMatch {
+
+	var r GcrPayloadMatch
+
+	if !strings.Contains(payload.Digest, (string)(image.ImageName)) {
+		return r
+	}
+	r |= FlagPathMatched
+	for digest, tags := range image.Dmap {
+		fqin := ToFQIN(rc.Name, image.ImageName, digest)
+		// The payload.Digest field actually holds the FQIN.
```

Comment:
Ah I see - and I presume that's an artifact of the GCR pubsub format?  If so, let's definitely convert that notification into something that is easier to work with...  even if we just rename Digest to `FQIN`